Source/WebCore/Modules/mediasource/MediaSourceHandle.cpp

```diff
@@ -0,0 +1,148 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "MediaSourceHandle.h"
+
+#if ENABLE(MEDIA_SOURCE_IN_WORKERS)
+
+#include "MediaSource.h"
+#include "MediaSourcePrivate.h"
+#include "MediaSourcePrivateClient.h"
+#include <wtf/ThreadSafeRefCounted.h>
+#include <wtf/ThreadSafeWeakPtr.h>
+
+namespace WebCore {
+
+WTF_MAKE_ISO_ALLOCATED_IMPL(MediaSourceHandle);
+
+struct MediaSourceHandle::SharedPrivate final : ThreadSafeRefCounted<SharedPrivate> {
+    static Ref<SharedPrivate> create(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher) { return adoptRef(*new SharedPrivate(mediaSource, WTFMove(dispatcher))); }
+
+    void dispatch(MediaSourceHandle::TaskType&& task, bool forceRunInWorker = false) const
+    {
+        m_dispatcher(WTFMove(task), forceRunInWorker);
+    }
+
+    void setMediaSourcePrivate(MediaSourcePrivate& mediaSourcePrivate)
+    {
+        Locker locker { m_lock };
+        m_private = mediaSourcePrivate;
+    }
+
+    RefPtr<MediaSourcePrivate> mediaSourcePrivate() const
+    {
+        Locker locker { m_lock };
+        return m_private.get();
+    }
+
+    friend class MediaSourceHandle;
+    SharedPrivate(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher)
+        : m_client(mediaSource.client().get())
+        , m_isManaged(mediaSource.isManaged())
+        , m_dispatcher(WTFMove(dispatcher))
+    {
+    }
+    mutable Lock m_lock;
+    ThreadSafeWeakPtr<MediaSourcePrivate> m_private WTF_GUARDED_BY_LOCK(m_lock);
+    std::atomic<bool> m_hasEverBeenAssignedAsSrcObject { false };
+    ThreadSafeWeakPtr<MediaSourcePrivateClient> m_client;
+    const bool m_isManaged;
+    const MediaSourceHandle::DispatcherType m_dispatcher;
+};
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(MediaSource& mediaSource, MediaSourceHandle::DispatcherType&& dispatcher)
+{
+    return adoptRef(*new MediaSourceHandle(mediaSource, WTFMove(dispatcher)));
+}
+
+Ref<MediaSourceHandle> MediaSourceHandle::create(Ref<MediaSourceHandle>&& other)
```

Comment:
This is difficult to read and follow.
I see why it is there but I feel like it would just be better to use the detached handle as a Ref<MediaSourceHandlePrivate> and create a MediaSourceHandle from Ref<MediaSourceHandlePrivate> directly.